### ISSUE TITLE
ImapClientDemo: InvalidOperationException: "Lock the SyncRoot property" on `folder.FetchAsync()`

### ISSUE THREAD

**Issue by schittli** [jstedfast/MailKit#1485]
Hello

**Thank you very much for sharing your great work!**, I try to use it for a non-profit project.

I tried pretty hard to find this bug, but because I do not know the concepts, I failed 😞 

**Note**: If it helps, I can send you the IMAP Server login, so that you can test it yourself.

**Describe the bug**
In ImapClientDemo, I click to a IMAP Mailbox folder to fetch the messages and then, I get the exception:
```
System.InvalidOperationException: 
'The ImapClient is currently busy processing a command in another thread. 
Lock the SyncRoot property to properly synchronize your threads.'
```

The exception happens on calling this line: [Permalink to MessageList.cs, Line 62](https://github.com/jstedfast/MailKit/blob/9f31a9c670d105b87f8903061dc88616aef75cdf/samples/ImapClientDemo/ImapClientDemo/MessageList.cs#L62)
```csharp
if (folder.Count > 0) {
  var summaries = await folder.FetchAsync (0, -1, request);
```

**Platform:**
 - OS: Windows
 - .NET Framework: 4.8.4515.0
 - .NET Runtime: [e.g. CoreCLR, Mono] ? I'm pretty sure it is not Mono
 - MailKit Version: Today, I've used:  
  `git clone https://github.com/jstedfast/MailKit.git`
   `git submodule update --init --recursive`

**Exception**
Exception Message:
```
System.InvalidOperationException: 
'The ImapClient is currently busy processing a command in another thread. 
Lock the SyncRoot property to properly synchronize your threads.'
```
StackTrace:
```csharp
System.InvalidOperationException
  HResult=0x80131509
  Message=The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads.
  Source=MailKit
  StackTrace:
   at MailKit.Net.Imap.ImapEngine.<IterateAsync>d__190.MoveNext() in D:\src\MailKit\MailKit\Net\Imap\ImapEngine.cs:line 2299

- $exception	{"The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads."}	System.InvalidOperationException
+ Data	{System.Collections.ListDictionaryInternal}	System.Collections.IDictionary {System.Collections.ListDictionaryInternal}
    HResult	-2146233079	int
    HelpLink	null	string
    IPForWatsonBuckets	0x0af623bb	System.UIntPtr
+ InnerException	null	System.Exception
    IsTransient	false	bool
    Message	"The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads."	string
    RemoteStackTrace	null	string
    Source	"MailKit"	string
    StackTrace	"   at MailKit.Net.Imap.ImapEngine.<IterateAsync>d__190.MoveNext() in D:\\src\\MailKit\\MailKit\\Net\\Imap\\ImapEngine.cs:line 2299"	string
+ TargetSite	{Void MoveNext()}	System.Reflection.MethodBase {System.Reflection.RuntimeMethodInfo}
    WatsonBuckets	null	object
    _HResult	-2146233079	int
    _className	null	string
+ _data	{System.Collections.ListDictionaryInternal}	System.Collections.IDictionary {System.Collections.ListDictionaryInternal}
    _dynamicMethods	null	object
    _exceptionMethod	null	System.Reflection.MethodBase
    _exceptionMethodString	null	string
    _helpURL	null	string
+ _innerException	null	System.Exception
    _ipForWatsonBuckets	0x0af623bb	System.UIntPtr
    _message	"The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads."	string
    _remoteStackIndex	0	int
    _remoteStackTraceString	null	string
+ _safeSerializationManager	{System.Runtime.Serialization.SafeSerializationManager}	System.Runtime.Serialization.SafeSerializationManager
    _source	"MailKit"	string
+ _stackTrace	{sbyte[24]}	object {sbyte[]}
    _stackTraceString	null	string
    _watsonBuckets	null	object
    _xcode	-532462766	int
    _xptrs	0x00000000	System.IntPtr
```

**To Reproduce**
I am doing:

1. Start ImapClientDemo and Login
2. I get the correct folder structure 😃:
![image](https://user-images.githubusercontent.com/8282673/206874213-d293bfd2-cecb-4fb2-bd5c-549bfdefacdb.png)
3. I click to the folder `90 Erledigt (2)`
4. On this line: [Permalink to MessageList.cs, Line 62](https://github.com/jstedfast/MailKit/blob/9f31a9c670d105b87f8903061dc88616aef75cdf/samples/ImapClientDemo/ImapClientDemo/MessageList.cs#L62)
```csharp
if (folder.Count > 0) {
  var summaries = await folder.FetchAsync (0, -1, request);
```

5. I get the exception

**Expected behavior**
It should fetch the messages 😃 

**Protocol Logs**
Here is the imap logfile: [imap.txt](https://github.com/jstedfast/MailKit/files/10200879/imap.txt)
_(The domain name is anonymized so that bots don't try to login)_

Kind regards,
Thomas

**Comment by jstedfast** [comment]
Thanks, I'll try to look into this. Seems like something must be causing 2+ threads to be making IMAP queries at the same time.

**Comment by ulued** [comment]
I have a simlilar issue calling GetMessageAsync from a folder. In order to process a large volume of messages I use various Threads calling differnt uids.  Maybe you can look that up also as it seems to be a similar issue related to threaded access. Locking Syncroot does not change anything.

System.InvalidOperationException: The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads.
   bei MailKit.Net.Imap.ImapEngine.<IterateAsync>d__190.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei MailKit.Net.Imap.ImapEngine.<RunAsync>d__191.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei MailKit.Net.Imap.ImapFolder.<GetMessageAsync>d__229.MoveNext()
   --- Ende der internen Ausnahmestapelüberwachung ---
   bei System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   bei windream.Imap.Mailprocessing.IMAPMailboxProcessorImapAsync.<IterateMailsAsync>d__25.MoveNext()
---> (Interne Ausnahme #0) System.InvalidOperationException: The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads.
   bei MailKit.Net.Imap.ImapEngine.<IterateAsync>d__190.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei MailKit.Net.Imap.ImapEngine.<RunAsync>d__191.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei MailKit.Net.Imap.ImapFolder.<GetMessageAsync>d__229.MoveNext()<---

**Comment by jstedfast** [comment]
@ulued This is a bug in the ImapClientDemo and not in MailKit itself.

The ImapClientDemo code is trying to fetch message summaries from a folder while also trying to query another folder for the number of unread messages.

Likely, if you are experiencing this same problem in your own app, your code is making the same type of mistake.

**Comment by ulued** [comment]
I have one folder (inbox) opended by await ... OpenFolderAsync. Next I get the UniqueIds of the messages of the one and only folder to a blocking collection. The first thread accessing a mail by an element of the blocking collection from this folder succeeds. The others might fail. I think the problem is in await folder.GetMessageAsync  which is not threadsafe for this kind of processing. Does MailKit support this kind of processing? I looks to me like await folder.GetMessageAsync is not awaited, if there is another thread doing the same call at more or less the same time to the same folder, or otherwise it would have succeeded. 

If I use my own lock 
lock(myLock)
{
    messasge = folder.GetMessageAsync().Result;
}
it works better, but still some calls might fail.  But I think await folder.GetMessageAsync  should do that.
Using just one thread to process the messages there are no problems

I can do this kind of processing in the MS Graph API without problems.. 
This is not be related to the ImapClientDemo - so open a separate issue for it?

**Comment by jstedfast** [comment]
@ulued no, don't bother opening an issue because the problem is in your code, not MailKit.

It sounds like you are trying to call `await folder.GetMessageAsync()` from multiple threads, but this is exactly why your code is getting the exception. This is NOT supported and can't be.

> I can do this kind of processing in the MS Graph API without problems..

That's because the GraphAPI client will open multiple HttpClient connections to make parallel requests, but MailKit can't easily do that because the IMAP protocol is not stateless.

**Comment by ulued** [comment]
Thanks, that explanation did help

**Comment by danzuep** [comment]
They way I've [fixed](https://github.com/danzuep/MailKitSimplified/issues/11) this [for my idle client](https://github.com/danzuep/MailKitSimplified/blob/main/source/MailKitSimplified.Receiver/Services/MailFolderMonitor.cs#L297) is by cloning the ImapClient and using multiple clients in parallel. This also means that any future calls to the IMessageSummary.Folder don't cause any issues with the idle client.

**Comment by ulued** [comment]
Thanks @danzuep  for pointing me to your solution. Jeremy already gave me the hint of using various connections in threads what I did.
